flake.nix

```diff
@@ -0,0 +1,90 @@
+{
+  description = "A static verifier for Rust, based on the Viper verification infrastructure.";
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
+    naersk.url = "github:nmattia/naersk";
+    rust-overlay.url = "github:oxalica/rust-overlay";
+    utils.url = "github:numtide/flake-utils";
+  };
+
+  outputs = { self, nixpkgs, naersk, rust-overlay, utils }:
+    utils.lib.eachDefaultSystem (system:
+      let
+        pkgs = import nixpkgs {
+          inherit system;
+          overlays = [ rust-overlay.overlay ];
+        };
+
+        rust = pkgs.rust-bin.fromRustupToolchainFile ./rust-toolchain;
+
+        naersk-lib = naersk.lib."${system}".override {
+          cargo = rust;
+          rustc = rust;
+        };
+
+        prusti-version = "${self.tag or "${self.lastModifiedDate}.${self.shortRev or "dirty"}"}";
+      in rec {
+        packages.prusti = naersk-lib.buildPackage {
+          name = "prusti";
+          version = "${prusti-version}";
+          root = ./.;
+          buildInputs = [
+            pkgs.pkg-config
+            pkgs.wget
+            pkgs.gcc
+            pkgs.openssl
+            pkgs.jdk11
+            packages.viper
+            packages.ow2_asm
+          ];
+          nativeBuildInputs = [
+            pkgs.makeWrapper
+          ];
+
+          override = _: {
+            preBuild = ''
+              export LD_LIBRARY_PATH="${pkgs.jdk11}/lib/openjdk/lib/server"
+              export VIPER_HOME="${packages.viper}/backends"
+              export Z3_EXE="${packages.viper}/z3/bin/z3"
+              export ASM_JAR="${packages.ow2_asm}/asm.jar"
+            '';
+            postInstall = ''
+              for f in $(find $out/bin/ $out/libexec/ -type f -executable); do
+                wrapProgram $f \
+                  --set PRUSTI_SYSROOT "${rust}" \
+                  --set JAVA_HOME "${pkgs.jdk11}/lib/openjdk" \
+                  --set LD_LIBRARY_PATH "${pkgs.jdk11}/lib/openjdk/lib/server" \
+                  --set VIPER_HOME "${packages.viper}/backends" \
+                  --set Z3_EXE "${packages.viper}/z3/bin/z3"
+              done
+            '';
+          };
+        };
+
+        packages.viper = pkgs.fetchzip {
+          name = "viper";
+          url = "https://viper.ethz.ch/downloads/ViperToolsNightlyLinux.zip";
+          stripRoot = false;
+          hash = "sha256-oV9VgGmzNrLW6n08vXU2rlV6S9Eca3U0L+gNbVxKhEE=";
```

Comment:
`https://viper.ethz.ch/downloads/ViperToolsNightlyLinux.zip` is overwritten often. `http://viper.ethz.ch/downloads/ViperToolsLastReleaseLinux.zip` might a bit more stable, but I haven't tried it. We are still working on a way to refer to an exact old nightly version of Viper.

Reply:
A possible solution suggested by @vakaras in Zulip (https://prusti.zulipchat.com/#narrow/stream/236614-general/topic/Building.20Prusti.20on.20NixOS/near/229701863), at least for NixOS, would be to build Viper from source using a separate Nix flake in its repository, which I started working on as well (will reference the relevant pull request when it's ready). It would then be possible to pin the dependency to a specific commit (done automatically by `flake.lock`, in fact), which can be updated as necessary by running `nix flake update`. The only caveat with this approach, as of now, would be the added time required to compile Viper in addition to Prusti. This should be resolved once a cache is hooked up, where it would be possible to have a workflow automatically compiling binaries of both Prusti and Viper on each commit to master, which will be made available to users by the Nix caching system.